boards: shields: Add support for Adafruit AW9523 board

soburi opened the pull request:
Add support for Adafruit AW9523 GPIO Expander/LED Controller

I have checked this PR with nulceo_h503rb using the equivalent board as a circuit (https://www.waveshare.com/wiki/AW9523B_IO_Expansion_Board) to Adafruit AW9523.

soburi committed:
drivers: gpio: Add support for Awinic AW9523B GPIO controller

Add support for GPIO controller feature of AW9523B.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>
Co-authored-by: Benjamin Cabé <kartben@gmail.com>

soburi committed:
boards: shields: Add support for Adafruit AW9523 board

Add support for Adafruit AW9523 GPIO Expander/LED Controller

Co-authored-by: Benjamin Cabé <kartben@gmail.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: drivers: gpio: gpio_api_1pin: Add AW9523B test configuration

Add test configuration for Awinic AW9523B.

This device is a GPIO Expander, and no boards currently implement it.
Define the connection that connects it via the Arduino header and
define a fixture to enable it.

This has been confirmed with nucleo_h503rb.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: drivers: gpio: gpio_basic_api: Add AW9523B test configuration

Add test configuration for Awinic AW9523B.

This device is a GPIO Expander, and no boards currently implement it.
Define the connection that connects it via the Arduino header and
define a fixture to enable it.

This has been confirmed with nucleo_h503rb.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: drivers: build_all: gpio: Add config for AW9523B

Add test configuration for "awinic,aw9523b".

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

kartben commented on boards/shields/adafruit_aw9523/adafruit_aw9523.overlay:
```diff
@@ -0,0 +1,26 @@
+ */
+
+/*
+ * This shield has the STEMMA-QT/Qwiic connector, but devicetree defines it as
+ * an electrically compatible Grove connector for convenience.
```
I appreciate you taking the time to document this "convenience shortcut" you took but I think you would actually need to define a proper connector + conventional aliases for Qwiic since otherwise this will be problematic for people trying to use this shield on board that has both types of connectors IMO

kartben commented on boards/shields/adafruit_aw9523/doc/adafruit_aw9523.jpg:
Use webp format please. Bonus points if you can find a version with transparent background (not mandatory, but nice to have)

soburi replied:
I changed the format to webp. I think the photos on the developer's page have the least copyright issues to quote, but unfortunately there were no photos with a white background here. Therefore, I did not change this.

kartben commented on boards/shields/adafruit_aw9523/doc/index.rst:
```diff
@@ -0,0 +1,66 @@
+
+   Adafruit AW9523 (Credit: Adafruit)
+
+.. note::
+   This shield has the STEMMA-QT/Qwiic connector, but devicetree defines it as
+   an electrically compatible Grove connector.
+
```
Drop when new proper connector will be adopted

kartben commented on boards/shields/adafruit_aw9523/doc/index.rst:
```diff
@@ -0,0 +1,66 @@
+Requirements
+************
+
+This shield can only be used with a board which provides a configuration for
+STEMMA QT/Qwiic or Grove.
```
```suggestion
This shield can only be used with a board which provides a configuration for
STEMMA QT/Qwiic.
```

soburi replied:
Fixed it.

kartben commented on boards/shields/adafruit_aw9523/doc/index.rst:
```diff
@@ -0,0 +1,66 @@
+
+This shield can only be used with a board which provides a configuration for
+STEMMA QT/Qwiic or Grove.
+defines node aliases for SPI and GPIO interfaces (see :ref:`shields` for more details).
```
What is this supposed to mean? Maybe a copy paste error?

kartben reviewed (changes requested):
Haven't reviewed the driver yet

soburi committed:
dts: bindings: gpio: Add STEMMA-QT connector

Add a definition of a STEMMA QT connector to which I2C
devices can be connected.

Since STEMMA-QT does not specify uses other than I2C,
this is just a formal definition.
(It may be helpful when using this connector as a GPIO,
 which is not defined in STEMMA-QT)

Maybe in usual need to define i2c alias such like as,

```
stemma_qt_i2c: &i2c0 { }
```

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>
Co-authored-by: Benjamin Cabé <kartben@gmail.com>

nordicjm commented on drivers/mfd/Kconfig.aw9523b:
```diff
@@ -0,0 +1,17 @@
+	  Enable the Awinic AW9523B GPIO/LED controller device driver
+
+config MFD_AW9523B_INIT_PRIORITY
+	int "nPM1300 MFD initialization priority"
```
wrong comment

soburi replied:
Thank you for pointing that out. This is embarrassing.
Fixed it.

nordicjm commented on boards/shields/adafruit_aw9523/doc/adafruit_aw9523.webp:
Could you delete the background from this (doesn't have to be transparent but remove the outer parts that are not useful for this imaging and making the filesize far larger), then run it through https://tinypng.com/

soburi replied:
Due to my circumstances, this is difficult for me to deal with. It passed the CI (<100kb), so I would appreciate it if you could accept it as is.

Background:

My country (Japan) has strict regulations about modifying other people's copyrighted work. If I modify the background, it does not fall under the category of quoting and requires permission.
There are differing opinions on whether copyright law should be treated on a territorial basis, which can create a worrying situation from a compliance perspective.

For these reasons, my only options are to get this approved or submit this commit without including the image.
Another idea is to take a new photo (it will not be beautiful!) to use here.

nordicjm replied:
No worries then, approved

soburi replied:
Thanks!

kartben commented on dts/bindings/gpio/awinic,aw9523b-gpio.yaml:
```diff
@@ -0,0 +1,36 @@
+    type: phandle-array
+    description: |
+      Set GPIO connected to the controller RSTN pin.
+      This is used to controlling resets via GPIO.
```
```suggestion
      This is used for controlling resets via GPIO.
```

soburi replied:
I applied your suggestion.

kartben commented on dts/bindings/gpio/stemma-qt-connector.yaml:
```diff
@@ -0,0 +1,21 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    STEMMA QT is a 1.0mm-JST-SH connector that
+    externalized the I2C bus.
```
```suggestion
    STEMMA QT is a 1.0mm-JST-SH connector for I2C devices.
```

soburi replied:
ditto

kartben commented on dts/bindings/mfd/awinic,aw9523b.yaml:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2024 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+description: Awinic AW9523B
```
```suggestion
description: Awinic AW9523B 16 multi-function LED driver and GPIO controller
```

soburi replied:
ditto

kartben commented on boards/shields/adafruit_aw9523/doc/index.rst:
```diff
@@ -0,0 +1,49 @@
+| RST (Pad on board)       | AW9523B Reset pin [2]_                   |
++--------------------------+------------------------------------------+
+
+.. [1] To receive interruption, connect the INT pin to the SoC's GPIO and set the connected
```
```suggestion
.. [1] To receive interrupts, connect the INT pin to the SoC's GPIO and set the connected
```

soburi replied:
ditto

kartben reviewed (changes requested):
minor comments - thanks!

soburi commented:
@henrikbrixandersen @mnkp 

> re-assigning to GPIO collaborators as it would be great to get this bless from a GPIO driver standpoint, as this is a much more substantial part of the PR than the shield bits :)

It would be helpful if you could also take a look at it from the perspective of whether the test can be easily applied to other GPIO Expanders.

kartben commented:
This might not get a review from GPIO collaborators after all. Maybe @fabiobaltieri will have some cycles to have a look, if it doesn't get merged before.

fabiobaltieri commented on drivers/gpio/Kconfig.aw9523b:
```diff
@@ -0,0 +1,18 @@
+	help
+	  Enable the AW9523B GPIO controller.
+
+config GPIO_AW9523B_INIT_PRIORITY
+	int "Init priority"
+	default 80
+	depends on GPIO_AW9523B
+	help
+	  AW9523B GPIO driver initialization priority.
```
can you try and ditch this and just use `CONFIG_MFD_INIT_PRIORITY`? Both here and on the mfd driver. Then subpriorities will control sequencing correctly and we have two less priority options kicking around.

soburi replied:
Thank you for your suggestion. This is the ideal solution.

fabiobaltieri commented on drivers/mfd/Kconfig.aw9523b:
```diff
@@ -0,0 +1,17 @@
+	help
+	  Enable the Awinic AW9523B GPIO/LED controller device driver
+
+config MFD_AW9523B_INIT_PRIORITY
+	int "Awinic AW9523B GPIO/LED controller initialization priority"
+	default 60
+	depends on MFD_AW9523B
+	help
+	  Initialization priority for AW9523B.
```
see my comment above, I think you can drop this

soburi replied:
ditto.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+	}
+
+	/* Either INPUT or OUTPUT must be set */
+	if ((!!(flags & GPIO_INPUT) ^ !!(flags & GPIO_OUTPUT)) == 0) {
```
that's a funny way of doing it, how about

`if ((flags & GPIO_INPUT) != 0 && (flags & GPIO_OUTPUT) != 0)`? bit easier to follow

soburi replied:
Here, I do XOR operations. The equivariant equation is $`{\displaystyle P\veebar Q=(P\lor Q)\land (\lnot P\lor \lnot Q)}`$
I corrected these lines following the equation.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+				goto on_error;
+			}
+
+			WRITE_BIT(data->prev_value, pin, !!((buf[1] << 8 | buf[0]) & BIT(pin)));
```
you can use `sys_get_le16` here

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+	}
+#endif
+
+	err = i2c_reg_update_byte_dt(&config->i2c, AW9523B_REG_OUTPUT0 + port, mask, out_high);
```
I'd do a define for `AW9523B_REG_OUTPUT0 + port` so you can do `AW9523B_REG_OUTPUT(port)`

soburi replied:
I fixed it.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+
+	k_sem_take(aw9523b_get_lock(config->mfd_dev), K_FOREVER);
+
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 1);
+	if (!err) {
+		err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT1, buf + 1, 1);
+	}
+
+	if (err) {
+		LOG_ERR("%s: Failed to read port status (%d)", dev->name, err);
```
this is quite hard to follow, could it be converted to something simpler? the usual `if (err) { goto unlock_and_return }` type of deal

soburi replied:
I've reorganized this entire function. I think this is getting better.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+	if (err) {
+		LOG_ERR("%s: Failed to read port status (%d)", dev->name, err);
+	} else if (mode != READ) {
+		gpio_port_value_t old_value = (buf[1] << 8 | buf[0]);
```
use `sys_get_le16`

soburi replied:
I fixed it.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+		}
+
+		if (new_value != old_value) {
+			buf[0] = (new_value >> 0) & 0xFF;
+			buf[1] = (new_value >> 8) & 0xFF;
```
`sys_put_le16` or something  :-)

soburi replied:
I fixed it and other similars.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+		if (new_value != old_value) {
+			buf[0] = (new_value >> 0) & 0xFF;
+			buf[1] = (new_value >> 8) & 0xFF;
+			err = i2c_burst_write_dt(&config->i2c, AW9523B_REG_OUTPUT0, buf, 2);
```
`err = i2c_burst_write_dt(&config->i2c, AW9523B_REG_OUTPUT0, buf, sizeof(buf));`

less likely to be copypasted somewhere incorrectly, I usually like all write to refer the source structure if possible, makes it obvious to see that it's right

soburi replied:
I fixed it and other similars.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+	k_sem_give(aw9523b_get_lock(config->mfd_dev));
+
+	if (err == 0 && mode == READ) {
+		*value = (buf[1] << 8 | buf[0]);
```
`sys_get_le16`, but also yeah it's hard to track what code path executes this, if you could refactor it to make all the breanches and exit paths obvious it'd be great

soburi replied:
I've reorganized this entire function. I think this is getting better.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+
+static int gpio_aw9523b_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	gpio_port_value_t zero = 0;
```
`const`? (don't think it makes any difference but why  not)

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+		LOG_ERR("%s: Failed to read INPUT0 %d", data->dev->name, err);
+	}
+
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT1, buf + 1, 1);
```
`&buf[1]`? personal preference, up to you, for some reasons I  find it more obvious when dereferencing intermediate buffer points

Also, the registeres are after one another, can't you just do `i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 2);`?

soburi replied:
The operation here is special; if the state of pins 0-7 changes, the interrupt will not be unlatched unless the address of INPUT0 is read, and there is the same relation between 8-15 pins and INPUT1.
At this point, we know that an interrupt has occurred, but we do not know on which side it occurred, so we need to read INPUT0 and INPUT1 independently to clear the interrupt.

fabiobaltieri replied:
got it, thanks for explaining, consider adding a comment for future reference

soburi replied:
OK, I'll address it.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+		LOG_ERR("%s: Failed to read INPUT1 %d", data->dev->name, err);
+	}
+
+	value = buf[1] << 8 | buf[0];
```
sys_get etc etc, you got the idea

soburi replied:
This was also addressed.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,474 @@
+			return err;
+		}
+
+		WRITE_BIT(data->prev_value, pin, !!((buf[1] << 8 | buf[0]) & BIT(pin)));
```
same as the similar one somewhere above

soburi replied:
This was also addressed.

fabiobaltieri commented:
@erwango can you take a look at the shield part of this?

soburi commented on dts/bindings/gpio/stemma-qt-connector.yaml:
```diff
@@ -0,0 +1,20 @@
+    Please see
+    https://learn.adafruit.com/introducing-adafruit-stemma-qt/stemma-qt-comparison#quick-comparison-3035247
+
+compatible: "stemma-qt-connector"
```
This point will likely cause some debate, but I think keeping the brands and trademarks as they are is best.
This is a kind of guarantee of interoperability, so I think following it would be less confusing.

As I wrote in the commit message, STEMMA QT is only for I2C, and this yaml file is merely a formal definition and doesn't practically have much meaning.
As for STEMMA QT, I think interoperability will be ensured by the naming convention for the labels assigned to i2c.

STEMMA QT and Qwiic have an inclusion relationship in their specifications, and STEMMA QT includes the Qwiic specifications.

In such cases, I think it would be better to create two aliases on the board side, that is,

```
stemma_qt_i2c: &i2c0 { }
qwiic_qt_i2c: &i2c0 { }
```

like this.

In software terms, the signals are in the same order, so it would be easier to unify them, including Grove. Still, I think it would be better to use their official names to identify them and make clear their interoperability.

fabiobaltieri replied:
does not seem like doubling up is worth it, I mean a shield will be either stemmma/qwiic or grove or the other one but the stemma/qwiic ones interoperate, let's pick one and go with it, really either is good

or rather is either is bad, can't believe they did all this work to do the branding and didn't think about adding an interrupt line to the connector, anyway...

soburi replied:
Qwiic is 3.3v, and Stemma is 3V - 5V, so the difficulty is that they are only partially "interoperable". Strictly speaking, qwiic devices can be connected to stemma buses, but Stemma devices cannot be connected to qwiic buses. Also, by definition, a 5V stemma bus can be considered, so there is no interoperability with the Qwiic device in this case. It is a challenging task to express that it will work properly if it can be connected (if it can be compiled in software terms).

Since this PR concerns only the device side, I think the current definition is fine.

> or rather is either is bad, can't believe they did all this work to do the branding and didn't think about adding an interrupt line to the connector, anyway...

This is it!

fabiobaltieri commented on dts/bindings/gpio/stemma-qt-connector.yaml:
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    STEMMA QT is a 1.0mm-JST-SH connector for I2C devices.
```
did you actually change anything here? still look the same

fabiobaltieri commented on dts/bindings/gpio/stemma-qt-connector.yaml:
```diff
@@ -0,0 +1,20 @@
+
+        0  SCL
+        1  SDA
+        -  VCC (3-5V)
+        -  GND
```
why did you skip the pin numbers for 2 and 3?

soburi replied:
This follows the same notation as the grove.
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/gpio/grove-header.yaml#L9-L12

Other connectors also use this notation.

fabiobaltieri replied:
Oh it's the pin id in the nexus node, of course! Ok nevermind sorry for the noise.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,504 @@
+
+	err = i2c_reg_update_byte_dt(&config->i2c, AW9523B_REG_INT(port), mask, n_int_en);
+	if (err) {
+		LOG_ERR("%s: Failed to configure pin interruption (%d)", dev->name, err);
```
you sure you want to go a head and keep running the rest of the code if this fails? seems a bit pointless, the bus is likely gone for good if we get here, and even if the following call works, what good does it do for us? may as well `return err` and `return 0` at the end

soburi replied:
As you pointed out, there was a route just below this that could cause the semaphore release to leak, so we addressed that as well.

fabiobaltieri commented on drivers/gpio/gpio_aw9523b.c:
```diff
@@ -0,0 +1,504 @@
+#define GPIO_AW9523B_DEFINE(inst)                                                                  \
+	static struct gpio_aw9523b_data gpio_aw9523b_data##inst;                                   \
+                                                                                                   \
+	IF_ENABLED(DT_INST_PROP_HAS_IDX(inst, int_gpios, 0), (                                     \
+		static void gpio_aw9523b_int_cb##inst(const struct device *dev,                    \
+							     struct gpio_callback *cb,             \
+							     gpio_port_pins_t pins)                \
+		{                                                                                  \
+			k_work_submit(&gpio_aw9523b_data##inst.intr_worker);                       \
+		}                                                                                  \
+	))                                                                                         \
```
I don't think you need to generate this, can you do

```c
static __maybe_unused void gpio_aw9523b_int_handler(const struct device *gpio_dev,
                                   struct gpio_callback *cb,
                                   uint32_t pins)
{       
        struct gpio_aw9523b_data *data = CONTAINER_OF(
                        cb, struct gpio_aw9523b_data, gpio_callback); 
                
        k_work_submit(&data->intr_worker);
}
```
instead? (copied it from input_paw32xx.c)

soburi replied:
I worried about whether __maybe_unused can be used. It is standardized in C23, not C99.
If I can use it, it is better to use it.

fabiobaltieri replied:
It's pretty widely used in the code base, I wouldn't worry about it.

soburi replied:
I got it. I applied your suggestion.

fabiobaltieri replied:
ok I see you added it on other callbacks but would you not use it to replace `gpio_aw9523b_int_cb##inst` with a common one as well?

soburi replied:
Ah, I misunderstood what you said.
This is better. I've addressed it.

fabiobaltieri reviewed (commented):
good stuff, few more comments

fabiobaltieri commented on dts/bindings/gpio/stemma-qt-connector.yaml:
```diff
@@ -2,7 +2,7 @@
 
 description: |
-    STEMMA QT is a 1.0mm-JST-SH connector for I2C devices.
+    STEMMA QT is a 4-pin JST-SH connector for I2C devices.
```
almost there, one more commit back :-)
